Add ResourceSlices field to kubectl describe node

ArangoGutierrez opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Adds a **ResourceSlices** section to `kubectl describe node`

#### Which issue(s) this PR fixes:

Fixes #XXXXXX <!-- Substitute with the tracking issue ID -->

#### Does this PR introduce a user-facing change?

```release-note
kubectl: `kubectl describe node` now lists aggregated **ResourceSlices** when the `ResourceSlice` API is present, detailing slice name, driver, and pool.
```

ArangoGutierrez commented:
/sig-node
/sig-cli

ArangoGutierrez commented:
/cc @pohly @klueska

kannon92 commented:
/wg device-management

ArangoGutierrez commented:
/test pull-kubernetes-integration

mortent commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -51,6 +51,7 @@ import (
 	policyv1 "k8s.io/api/policy/v1"
 	policyv1beta1 "k8s.io/api/policy/v1beta1"
 	rbacv1 "k8s.io/api/rbac/v1"
+	resourcev1beta2 "k8s.io/api/resource/v1beta2"
```
The DRA controllers still use the `v1beta1` API, so we should probably use that here too. It is possible to use DRA without enabling the `v1beta2` API. We should update this when we update the controllers.

ArangoGutierrez replied:
We need to aim for `beta2` as it introduces the `spec.NodeName` as a server side filter. The DRA repo `master` branch is already using this version, so is just a matter of time for it to cut a version supporting `v1beta2` .
ON `v1beta1` we would have to do the filtering on the client side

pohly replied:
> We need to aim for beta2 as it introduces the spec.NodeName as a server side filter.

That's not true: all API versions support that filter.

ArangoGutierrez replied:
My bad, then I'll move back to v1beta1 for this PR

pohly replied:
Let's merge this with v1beta1 for consistency. We will then convert to v1 with the rest of the code during GA graduation.

ArangoGutierrez replied:
edited to v1beta1

mortent commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -3762,14 +3763,24 @@ func (d *NodeDescriber) Describe(namespace, name string, describerSettings Descr
 
-	return describeNode(node, nodeNonTerminatedPodsList, events, canViewPods, &LeaseDescriber{d})
+	var resourceSliceList *resourcev1beta2.ResourceSliceList
+	resourceSliceList, err = d.ResourceV1beta2().ResourceSlices().List(context.TODO(), metav1.ListOptions{
```
So this would only show the ResourceSlices that are associated with the node through the `spec.nodeName` field. But other ResourceSlices can also be available to the node, either through the `spec.allNodes` or `spec.nodeSelector` fields. And with the Partitionable Devices feature, it is possible to associate individual devices with nodes, not just whole ResourceSlices. How should we handle these other situations?

pohly replied:
Let's merge this with v1beta1 for consistency. We will then convert to v1 with the rest of the code during GA graduation.

ArangoGutierrez replied:
I did some changes with some client side filtering, now it looks like
```bash
ResourceSlices:
Name                    Driver              Pool            Association
----                    ------              ----            -----------
slice-1                 example.com/gpu     gpu-pool        Direct node match
slice-2                 example.com/fpga    fpga-pool       Available to all nodes
slice-3                 example.com/gpu     gpu-pool-2      Via node selector
slice-4                 example.com/nic     nic-pool        Per-device selection
  Devices:
    - gpu-1 (Node: worker-node-1)
    - gpu-2 (Node: worker-node-1)
  ```

johnbelamaric replied:
I don't think we should show any ResourceSlices other than those that are directly associated with this node. There are too many others possible, it's not what we would want in a Node view.

mortent replied:
So I don't think we should show individual devices, as that seems too much detail.

But I'm not quite sure what is the use-case for showing only the ResourceSlices that are associated with the node through the NodeName field. It shows "some of the ResourceSlices that contains devices that can be used from this node", so it is an incomplete view of the state in the cluster.
Although maybe in practice we expect all node-local devices to always be published in ResourceSlices using the NodeName field, so it will show a list of ResourceSlices with devices exclusive to this node?

ArangoGutierrez replied:
That means `spec.NodeName = THIS_NODE` ONLY, and if the user wants more info they can inspect the ResourceSlices in the Cluster. Meaning `kubectl describe node` won't show/filter for `spec.NodeSelector` or `spec.AllNodes` as they could show ResourceSlices that are shared with many other nodes. 

Is that correct?

johnbelamaric replied:
That's what I think. Simply because it's impractical, and not really what the user is typically looking for if they are describing nodes, in my opinion. We don't know if people are going to use the AllNodes for example, for a network accessible set of volumes or even abstract resources (e.g., licenses). There could just be so many slices, and many may be irrelevant to what the user is actually looking for.

We may want to qualify what the list of resource slices is, with something like "ResourceSlices available only on this node" instead of just "ResourceSlices". 

I agree listing individual devices is probably too detailed, but a count would be great. If we could calculate a count of used devices out of a pool that would be nice, but it's a bit tedious - you would look at all pods, all resource claims used by those pods, then match the pools and number of devices used. This would also get a little weird because some devices can be shared. And once we have consumable capacity, just because it's used doesn't mean it's unavailable. I guess just saying "In use" and a count could be useful though.

ArangoGutierrez replied:
Before jumping to address these comments as PR author, I would like to work on this thread so we all agree on what to show on the describe node.

Thoughts @mortent @pohly @klueska  ?

johnbelamaric replied:
Showing pools, not slices, might be better. It would display a lot less data. You could do driver/pool/#slices/#devices.

You still have to read all slices in the cluster to do this, though, which seems painful. I was hoping we could avoid that, because it means the time to describe a node is O(number of slices in the cluster), rather than O(1). For a cluster with 10,000 GPU nodes, that's going to be super slow. Not sure our API query mechanisms are good enough to get around that though, without adding a label for indexing - should we do that?

johnbelamaric replied:
Even if we did that it wouldn't help for non-node-local ones though

pohly replied:
> Not sure our API query mechanisms are good enough to get around that though, without adding a label for indexing

We have an indexed label on the node name in a ResourceSlice. But as you said, that's not useful in this context because it would only provide an incomplete view.

> For a cluster with 10,000 GPU nodes, that's going to be super slow.

Do we know how slow?

I am facing a similar challenge over in DRA device taints: I want to show the effect of a DeviceTaintRule without actually triggering the effect. `kubectl describe DeviceTaintRule` seems the best place for that, but it needs to read both slices and pods - see https://github.com/kubernetes/enhancements/pull/5512.

ArangoGutierrez commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -51,6 +51,7 @@ import (
 	policyv1 "k8s.io/api/policy/v1"
 	policyv1beta1 "k8s.io/api/policy/v1beta1"
 	rbacv1 "k8s.io/api/rbac/v1"
+	resourcev1beta1 "k8s.io/api/resource/v1beta1"
```
That change is pretty straightforward, as the API's I am using exist there as well.  DO you have any comments on the logic added by the patch? 

We might decide on beta1 or beta2 before merging the PR, but I want to have a deeper conversation around https://github.com/kubernetes/kubernetes/pull/131744#discussion_r2087404273

ArangoGutierrez replied:
And thanks by the way!

johnbelamaric commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -3861,13 +3873,77 @@ func describeNode(node *corev1.Node, nodeNonTerminatedPodsList *corev1.PodList,
+		// Determine association type
+		switch {
+		case slice.Spec.NodeName != "" && slice.Spec.NodeName == node.Name:
+			association = "Direct node match"
```
I think these are the ONLY ones we should show. We should filter this out earlier and not show any others. There can really be a lot of resources slices "available from" a node - think thousands. We shouldn't show the "could be used on this node" as the same as "can ONLY be used on this node". The latter are all we want in a per-node view.

To see the other things, users should look at the resource slices directly.

johnbelamaric commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -3861,13 +3873,77 @@ func describeNode(node *corev1.Node, nodeNonTerminatedPodsList *corev1.PodList,
+				slice.Spec.Pool.Name,
+				association)
+
+			// If per-device selection is enabled, show assigned devices
```
Even these devices that are part of a larger resource slice and are per-node, we do not need to display here.

johnbelamaric commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -3861,13 +3873,77 @@ func describeNode(node *corev1.Node, nodeNonTerminatedPodsList *corev1.PodList,
+				slice.Name,
+				slice.Spec.Driver,
+				slice.Spec.Pool.Name,
+				association)
```
v1beta2 will have "Devices" rather than the Basic/Composite oneOf. So, I think you can display a count of the devices here too.

soltysh reviewed (approved):
this looks reasonable from sig-cli pov, I'll leave the final tag to John and/or Patrick

/approve

johnbelamaric commented:
I am concerned about the scale of this command on large clusters with many slices accessible from many nodes. What are the policies around changing describe output, should we find this is just "too much"?

ArangoGutierrez committed:
kubectl: Add ResourceSlices to describe node with O(1) query

Add a Node-Exclusive ResourceSlices section to `kubectl describe node`
that displays DRA ResourceSlices assigned exclusively to the node.

Key design decisions per reviewer feedback:

1. Uses indexed field selector (spec.nodeName) for O(1) query performance
   instead of fetching all slices cluster-wide. This is critical for
   large clusters with thousands of ResourceSlices.

2. Shows only node-exclusive slices (where spec.NodeName == thisNode).
   Does NOT show AllNodes, NodeSelector, or PerDeviceNodeSelection slices
   as these could number in the thousands and aren't node-specific.

3. Aggregates by driver/pool instead of listing individual slices:
   - Shows: Driver, Pool, Slices count, Devices count
   - More concise and useful output

4. Caps output at 10 pools with "...and X more pools" message,
   following the existing formatEndpointSlices pattern.

Example output:
  Node-Exclusive ResourceSlices:
    Driver              Pool        Slices  Devices
    ------              ----        ------  -------
    nvidia.com/gpu      gpu-pool    2       8
    intel.com/qat       qat-pool    1       4

Signed-off-by: Carlos Eduardo Arango Gutierrez <eduardoa@nvidia.com>

pohly commented:
> Now uses the indexed field selector (spec.nodeName) for O(1) query performance instead of fetching all ResourceSlices cluster-wide

This misses slices where node affinity is set per-device (a new feature added together with partitionable devices).

> Node-Exclusive ResourceSlices:

Instead of "node-exclusive" I tended to call this "node-local".

This might be good enough for most use cases. We just need to make users aware of the limitations.

/cc @mortent

johnbelamaric commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -3582,6 +3594,58 @@ func describeNode(node *corev1.Node, nodeNonTerminatedPodsList *corev1.PodList,
+	}
+	sort.Strings(sortedKeys)
+
+	w.Write(LEVEL_0, "Node-Exclusive ResourceSlices:\n")
```
s/Exclusive/Local/

johnbelamaric reviewed (commented):
modulo small changes noted, LGTM. Thank you!

ArangoGutierrez commented:
/assign [seans3](https://github.com/seans3)

ArangoGutierrez commented:
/remove-lifecycle rotten

johnbelamaric commented:
/assign @seans3

ArangoGutierrez commented:
/retest-required

soltysh reviewed (approved):
/lgtm
/approve
